tpu-client/src/tpu_connection_cache.rs

```diff
@@ -0,0 +1,561 @@
+        tpu_connection::TpuConnection as BlockingTpuConnection,
+    },
+    indexmap::map::IndexMap,
+    rand::{thread_rng, Rng},
+    solana_measure::measure::Measure,
+    solana_sdk::timing::AtomicInterval,
+    std::{
+        net::SocketAddr,
+        sync::{atomic::Ordering, Arc, RwLock},
+    },
+    thiserror::Error,
+};
+
+// Should be non-zero
+pub(crate) static MAX_CONNECTIONS: usize = 1024;
+
+/// Used to decide whether the TPU and underlying connection cache should use
+/// QUIC connections.
+pub const DEFAULT_TPU_USE_QUIC: bool = true;
+
+/// Default TPU connection pool size per remote address
+pub const DEFAULT_TPU_CONNECTION_POOL_SIZE: usize = 4;
+
+pub const DEFAULT_TPU_ENABLE_UDP: bool = false;
+
+pub struct TpuConnectionCache<P: ConnectionPool> {
+    pub map: RwLock<IndexMap<SocketAddr, P>>,
+    pub stats: Arc<ConnectionCacheStats>,
+    pub last_stats: AtomicInterval,
+    pub connection_pool_size: usize,
+    pub tpu_config: P::TpuConfig,
+}
+
+impl<P: ConnectionPool> TpuConnectionCache<P> {
+    pub fn new(connection_pool_size: usize) -> Self {
+        Self {
+            connection_pool_size: 1.max(connection_pool_size), // The minimum pool size is 1.
+            ..Self::default()
+        }
+    }
+
+    /// Create a lazy connection object under the exclusive lock of the cache map if there is not
+    /// enough used connections in the connection pool for the specified address.
+    /// Returns CreateConnectionResult.
+    fn create_connection(
+        &self,
+        lock_timing_ms: &mut u64,
+        addr: &SocketAddr,
+    ) -> CreateConnectionResult<P::PoolTpuConnection> {
+        let mut get_connection_map_lock_measure = Measure::start("get_connection_map_lock_measure");
+        let mut map = self.map.write().unwrap();
+        get_connection_map_lock_measure.stop();
+        *lock_timing_ms = lock_timing_ms.saturating_add(get_connection_map_lock_measure.as_ms());
+        // Read again, as it is possible that between read lock dropped and the write lock acquired
+        // another thread could have setup the connection.
+
+        let should_create_connection = map
+            .get(addr)
+            .map(|pool| pool.need_new_connection(self.connection_pool_size))
+            .unwrap_or(true);
+
+        let (cache_hit, num_evictions, eviction_timing_ms) = if should_create_connection {
+            // evict a connection if the cache is reaching upper bounds
+            let mut num_evictions = 0;
+            let mut get_connection_cache_eviction_measure =
+                Measure::start("get_connection_cache_eviction_measure");
+            while map.len() >= MAX_CONNECTIONS {
+                let mut rng = thread_rng();
+                let n = rng.gen_range(0, MAX_CONNECTIONS);
+                map.swap_remove_index(n);
+                num_evictions += 1;
+            }
+            get_connection_cache_eviction_measure.stop();
+
+            map.entry(*addr)
```

Comment:
Very small nit: this is ever so slightly different from the original in that the original would clone an entry's Quic endpoint before starting the eviction, so we wouldn't have to recreate an endpoint on the off chance that Entry happened to be evicted. Not a huge deal, but would be nice to not have that happen if there's a reasonably simple way to do it... Maybe just cache the endpoint in the QuicConfig? Unless the the QuicConfig is mutated, it should always create an Endpoint with the same params

Reply:
The endpoint is actually stored on the QuicPool itself. I removed the Option, because it's clearly assumed to be Some on all connection creation: https://github.com/solana-labs/solana/blob/80c01734528c8a37c6c10a878547f5e4ee85cb0a/tpu-client/src/connection_cache.rs#L373

Reply:
Basically, the issue (and this is quite minor) is, what happens if that QuicPool happens to be one of the ones evicted? In the original code, we would clone the endpoint before doing the eviction.

Reply:
The QuicLazyInitializedEndpoint is lazily inittialized, it shouldn't be a huge deal to just have an Arc to it in both the QuicConfig. The QuicConfig can have an Option\<Arc\<QuicLazyInitializedEndpoint\>\> to "invalidate" its cached QuicLazyInitializedEndpoint when mutated. We can then emulated the previous behavior of not recreating the Endpoint if we happen to evict that pool for the address we're creating a connection for. If it's too much work, it's probably fine...

Reply:
I follow now, but would a better approach be to not evict the QuicPool we're about to use?

Reply:
Something like this:
```diff
--- a/tpu-client/src/tpu_connection_cache.rs
+++ b/tpu-client/src/tpu_connection_cache.rs
@@ -71,6 +71,11 @@ impl<P: ConnectionPool> TpuConnectionCache<P> {
             while map.len() >= MAX_CONNECTIONS {
                 let mut rng = thread_rng();
                 let n = rng.gen_range(0, MAX_CONNECTIONS);
+                if let Some(index) = map.get_index(addr) {
+                    if n == index {
+                        continue;
+                    }
+                }
                 map.swap_remove_index(n);
                 num_evictions += 1;
             }
```

Reply:
I _think_ that should probably be fine, though it might affect the random behavior of the cache for small numbers of addresses... I'm not an expert on cache strategies, but I don't think it should be huge deal...

Reply:
Okay, thought about both approaches overnight, and I'm going to go with the eviction diff. I think it is in the spirit of the intended eviction logic. Also, it will only affect a very small number of cases, when: QuicPool entry at addr exists, but has only 1-3 existing connections, and the random index happens to hit (1/1024+).
Even if addr is being requested here frequently, we could only hit the case 3 times before the first criterion is no longer met.